# Q1. Write a function that inputs a number and prints the multiplication table of that number

In [28]:
num = int(input(print("Enter number to get multiplication table")))

def multiply(num1):
    i = product = 1
    while i < 11:
        product = num1 * i
        print(str(num1) +"*" +str(i) + "=" + str(product))
        i += 1
        
print(multiply(num))

Enter number to get multiplication table
None11
11*1=11
11*2=22
11*3=33
11*4=44
11*5=55
11*6=66
11*7=77
11*8=88
11*9=99
11*10=110
None


# Q2. Write a program to print twin primes less than 1000. If two consecutive odd numbers are
both prime then they are known as twin primes

In [2]:
def isPrime(n):
    for i in range(2, n):
        if n%i == 0:
            return False
    return True

def generateTwins():
    for i in range(3, 1000, 2):
        j = i+2
        if(isPrime(i) and isPrime(j)):
            print("{0} & {1}".format(i,j))

        
generateTwins()                      

3 & 5
5 & 7
11 & 13
17 & 19
29 & 31
41 & 43
59 & 61
71 & 73
101 & 103
107 & 109
137 & 139
149 & 151
179 & 181
191 & 193
197 & 199
227 & 229
239 & 241
269 & 271
281 & 283
311 & 313
347 & 349
419 & 421
431 & 433
461 & 463
521 & 523
569 & 571
599 & 601
617 & 619
641 & 643
659 & 661
809 & 811
821 & 823
827 & 829
857 & 859
881 & 883


# Q3. Write a program to find out the prime factors of a number. Example: prime factors of 56 - 2, 2, 2, 7

In [5]:
import math

def primeFactors():
    
    n = int(input(print("Enter number to find prime factors")))
    
    lstfactors = []
    
    #for even numbers 
    while n%2 == 0:
        lstfactors.append(2)
        n = n/2
        
    #Every composite number has at least one prime factor less than or equal to square root of itself, so calculating odd
    for i in range(3,int(math.sqrt(n))+1,2): 
        while n % i== 0: 
            lstfactors.append(i) 
            n = n / i 
    #Condition if n is a prime number greater than 2
    if n > 2: 
        lstfactors.append(n) 
        
    return lstfactors

primeFactors()

Enter number to find prime factors
None990


[2, 3, 3, 5, 11]

# Q4. Write a program to implement these formulae of permutations and combinations. Number of permutations of n objects taken r at a time: p(n, r) = n! / (n-r)!. Number of combinations of n objects taken r at a time is: c(n, r) = n! / (r!*(n-r)!) = p(n,r) / r!

In [6]:
def factorial(num):
    return 1 if num == 1 else (num * factorial(num-1))

def perm(n,r):
    result = factorial(n)/factorial(n-r)
    return result
    
def comb(n,r):
    result = perm(n,r)/factorial(r)
    return result

num_n = int(input(print("Enter greater number n to find permutation and combination")))
num_r = int(input(print("Enter smaller number r to find permutation and combination")))    
print("Permutation of ("+ str(num_n) + ", "+ str(num_r) + ") is:" + str(perm(num_n, num_r)))
print("Combination of ("+ str(num_n) + ", "+ str(num_r) + ") is:" + str(comb(num_n, num_r)))

Enter greater number n to find permutation and combination
None4
Enter smaller number r to find permutation and combination
None3
Permutation of (4, 3) is:24.0
Combination of (4, 3) is:4.0


# Q5. Write a function that converts a decimal number to binary number

In [8]:
num = int(input(print("Enter number:")))

def dec2bin(num1):
    """Function to convert decimal number to binary number."""
    
    #Array of size num1 
    binNum = [0] * num1
    
    #initialize index to 0
    i = 0
    
    """storing binNum array with remainder of num1 
    after dividing by 2 & storing quotient in num1 again 
    then incrementing"""
    while(num1 >0):
        binNum[i] = num1 % 2
        num1 = int(num1/2)
        i+=1
        
        
    #printing binNum array in reverse order
    for j in range(i-1, -1, -1):
        print(binNum[j], end = '')
    
    
    
dec2bin(num)

Enter number:
None9
1001

# Q6. Write a function cubesum() that accepts an integer and returns the sum of the cubes of individual digits of that number. Use this function to make functions PrintArmstrong() and isArmstrong() to print Armstrong numbers and to find whether is an Armstrong number.

In [19]:
def cubesum(num1):
    dig = list(d for d in str(num1))
    sum1 = 0
    for i in dig:
        sum1 += int(i)**3
        
    return sum1

def PrintArmstrong(num1):
    print("Sum is:" + str(cubesum(num1)))

def isArmstrong(num1, num2):
    if num1 == num2:
        print(str(num1) + " is an Armstrong number.")
    else:
        print(str(num1) + " is not an Armstrong number.")

def acceptNumber():
    num = int(input(print("Enter integer:")))
    isArmstrong(num, cubesum(num))
    PrintArmstrong(num)
    
    
acceptNumber()

Enter integer:
None371
371 is an Armstrong number.
Sum is:371


# Q7. Write a function prodDigits() that inputs a number and returns the product of digits of that number.

In [1]:
def prodDigits(num1):
    dig = list(d for d in str(num1))
    prod = 1
    for i in dig:
        prod *= int(i)
    
    return prod

def acceptNumber():
    num = int(input(print("Enter integer:")))
    print("Product of digits of the integer is: " + str(prodDigits(num)))
    
acceptNumber()

Enter integer:
None345
Product of digits of the integer is: 60


# Q8. If all digits of a number n are multiplied by each other repeating with the product, the one digit number obtained at last is called the multiplicative digital root of n. The number of times digits need to be multiplied to reach one digit is called the multiplicative persistance of n. Example: 86 -> 48 -> 32 -> 6 (MDR 6, MPersistence 3), 341 -> 12->2 (MDR 2, MPersistence 2). Using the function prodDigits() of previous exercise write functions MDR() and MPersistence() that input a number and return its multiplicative digital root and multiplicative persistence respectively

In [1]:
def prodDigits(num):
    dig = list(d for d in str(num))
    prod = 1
    for i in dig:
        prod *= int(i)
    return prod

def MPersistence(num, counter):
    counter += 1
    return counter-1 if num<10 else MPersistence(prodDigits(num), counter)

def MDR(num):
    return num if num<10 else MDR(prodDigits(num))
        
def acceptNumber():
    num = int(input(print("Enter integer:")))
    counter = 0
    print("(MDR " + str(MDR(num)) + ", MPersistence " + str(MPersistence(num, counter)) + ")")
    
    
acceptNumber()

Enter integer:
None86
(MDR 6, MPersistence 3)


# Q9. Write a function sumPdivisors() that finds the sum of proper divisors of a number. Proper divisors of a number are those numbers by which the number is divisible, except the number itself. For example proper divisors of 36 are 1, 2, 3, 4, 6, 9, 18

In [1]:
def sumPdivisors():
    num = int(input(print("Enter the number:")))
    divs = list(i for i in range(1, num) if num%i == 0)
    print("Proper divisors of " + str(num) + " are: " + str(divs))
    sum1 = 0
    
    for i in divs:
        sum1 += i
    print("Sum of proper divisors of " + str(num) + " is: " + str(sum1))

sumPdivisors()

Enter the number:
None36
Proper divisors of 36 are: [1, 2, 3, 4, 6, 9, 12, 18]
Sum of proper divisors of 36 is: 55


# 10. A number is called perfect if the sum of proper divisors of that number is equal to the number. For example 28 is perfect number, since 1+2+4+7+14=28. Write a program to print all the perfect numbers in a given range

In [14]:
def sumDivisor(num):
    divs = list(i for i in range(1, num) if num%i == 0)
    sum1 = 0
    
    for i in divs:
        sum1 += i
    return sum1

def sumEqualsNum(range1):
    x = list(i for i in range(1,range1) if(i == sumDivisor(i)))
    print("List of Perfect numbers in range " + str(range1) + ": " + str(x))
    
def checkPerfectNum():
    num = int(input(print("Enter max range:")))
    sumEqualsNum(num)
    
checkPerfectNum()

Enter max range:
None1000
List of Perfect numbers in range 1000: [6, 28, 496]


# Q11. Two different numbers are called amicable numbers if the sum of the proper divisors of each is equal to the other number. 

# For example 220 and 284 are amicable numbers. 
# Sum of proper divisors of 220 = 1+2+4+5+10+11+20+22+44+55+110 = 284. 
# Sum of proper divisors of 284 = 1+2+4+71+142 = 220. 

# Write a function to print pairs of amicable numbers in a range.

In [30]:
def sumDivisor(num):
    divs = list(i for i in range(1, num) if num%i == 0)
    sum1 = 0
    
    for i in divs:
        sum1 += i
    return sum1
    
def findPairs():
    range1 = int(input(print("Enter max range:")))
    num1 = num2 = 0
    print("Pair of amicable numbers in range " + str(range1) + " is:")
    for i in range(1, range1):
        num1 = i
        num2 = sumDivisor(num1)
        if ((num1 < num2) and (sumDivisor(num2) == num1)):
            print("(" + str(num1) + ", " + str(num2) +")") 
    return

findPairs()

Enter max range:
None10000
Pair of amicable numbers in range 10000 is:
(220, 284)
(1184, 1210)
(2620, 2924)
(5020, 5564)
(6232, 6368)


# Q12. Write a program which can filter odd numbers in a list by using filter function

In [6]:
def oddNum(n):
    if n%2 == 1:
        return n

    
number_list = range(-10, 10)
odd_num_lst = list(filter(oddNum, number_list))
print(odd_num_lst)
    

[-9, -7, -5, -3, -1, 1, 3, 5, 7, 9]


# Q13. Write a program which can map() to make a list whose elements are cube of elements in a given list

In [10]:
def cube(num):
    return num ** 3

number_list = range(-5, 6)
cubed = list(map(cube, number_list))
print(cubed)

[-125, -64, -27, -8, -1, 0, 1, 8, 27, 64, 125]


# Q14. Write a program which can map() and filter() to make a list whose elements are cube of even number in a given list

In [13]:
def evenNum(n):
    if n%2 == 0:
        return n
    
def cube(num):
    return num ** 3
    
number_list = range(-10, 10)
even = list(filter(evenNum, number_list))
cubed_even = list(map(cube, even))
print(cubed_even)

[-1000, -512, -216, -64, -8, 8, 64, 216, 512]
